src/app/core/services/github.service.ts

```diff
@@ -495,32 +495,55 @@ export class GithubService {
-   * Said function will recursively query for all subsequent pages until a page that has less than 100 items is found,
-   * then return all queried pages in an array.
+   * Returns an observable that will continually emit the currently accumulated results, until a page that has less
+   * than 100 items is found, after which it performs a final emit with the full results array, and completes.
    *
    * @callback pluckEdges - A function that returns a list of edges in a ApolloQueryResult.
+   * @params query - The query to be performed.
+   * @params variables - The variables for the query.
    * @returns an async function that accepts a GraphQL query for paginated data and any additional variables to that query
    */
-  private withPagination<T>(pluckEdges: (results: ApolloQueryResult<T>) => Array<any>) {
-    return async (query: DocumentNode, variables: { [key: string]: any } = {}): Promise<Array<ApolloQueryResult<T>>> => {
-      const maxResultsCount = 100;
-      const cursor = variables.cursor || null;
-      const graphqlQuery = this.apollo.watchQuery<T>({ query, variables: { ...variables, cursor } });
-      return graphqlQuery.refetch().then(async (results: ApolloQueryResult<T>) => {
+  private withPagination<T>(
+    pluckEdges: (results: ApolloQueryResult<T>) => Array<any>,
+    query: DocumentNode,
+    variables: { [key: string]: any } = {}
+  ): Observable<ApolloQueryResult<T>[]> {
+    const maxResultsCount = 100;
+    const apollo = this.apollo;
+
+    let accumulatedResults: ApolloQueryResult<T>[] = [];
+    const behaviorSubject: BehaviorSubject<ApolloQueryResult<T>[]> = new BehaviorSubject(accumulatedResults);
+    const observable = behaviorSubject.asObservable();
+
+    async function queryWith(cursor: string): Promise<void> {
+      const graphqlQuery = apollo.watchQuery<T>({ query, variables: { ...variables, cursor } });
+
+      await graphqlQuery.refetch().then(async (results: ApolloQueryResult<T>) => {
         const intermediate = Array.isArray(results) ? results : [results];
         const edges = pluckEdges(results);
         const nextCursor = edges.length === 0 ? null : edges[edges.length - 1].cursor;
 
+        accumulatedResults = accumulatedResults.concat(intermediate);
+        behaviorSubject.next(accumulatedResults);
         if (edges.length < maxResultsCount || !nextCursor) {
-          return intermediate;
+          // No more queries to perform.
+          return;
         }
-        const nextResults = await this.withPagination<T>(pluckEdges)(query, {
-          ...variables,
-          cursor: nextCursor
-        });
-        return intermediate.concat(nextResults);
+
+        // Use a chain of await to ensure that all recursive queries are completed before `complete` is called.
+        await queryWith(nextCursor);
       });
+    }
+
+    const initialCursor = null;
+    const observer = async () => {
+      await queryWith(initialCursor);
+      behaviorSubject.next(accumulatedResults);
```

Comment:
```suggestion
```

Not neccessary, before `queryWith` ends, the last `accumulatedResults` is already updated

Reply:
Without the final `behaviorSubject.next(accumulatedResults)`, it seems like the final batch of PRs wouldn't actually be sent. However, I'm not entirely sure why this is the case :/

Reproducible by removing this line, going to `catcher-org/catcher`, and searching for PR `1203` (one of the newer PRs)

I tried making the following change, together with the other comment, but it's still facing the same issue

```
M src/app/core/services/github.service.ts
@@ -527,6 +527,7 @@ export class GithubService {
         behaviorSubject.next(accumulatedResults);
         if (edges.length < maxResultsCount || !nextCursor) {
           // No more queries to perform.
+          behaviorSubject.complete();
           return;
         }
 
@@ -535,14 +536,15 @@ export class GithubService {
       });
     }
 
-    const initialCursor = null;
-    const observer = async () => {
-      await queryWith(initialCursor);
-      behaviorSubject.next(accumulatedResults);
-      behaviorSubject.complete();
-    };
-
-    observer();
+    queryWith(null);
 
     return observable;
   }
```

Reply:
see [comment](https://github.com/CATcher-org/WATcher/pull/179#issuecomment-1654671675)